platforms/core-runtime/internal-instrumentation-processor/src/test/groovy/org/gradle/internal/instrumentation/processor/codegen/jvmbytecode/InterceptJvmCallsGeneratorTest.groovy

```diff
@@ -219,4 +220,45 @@ class InterceptJvmCallsGeneratorTest extends InstrumentationCodeGenTest {
             .generatedSourceFile(fqName(expectedJvmInterceptors))
             .containsElementsIn(expectedJvmInterceptors)
     }
+
+    def "should fail if we use the same class for instrumentation as is for property upgrades"() {
+        given:
+        def instrumentation = source """
+            package org.gradle.test;
+            import org.gradle.internal.instrumentation.api.annotations.*;
+            import org.gradle.internal.instrumentation.api.annotations.CallableKind.*;
+            import org.gradle.internal.instrumentation.api.annotations.ParameterKind.*;
+            import org.gradle.internal.classpath.declarations.*;
+            import org.gradle.api.*;
+
+            @SpecificJvmCallInterceptors(generatedClassName = "${JVM_BYTECODE_GENERATED_CLASS_NAME_FOR_PROPERTY_UPGRADES}_TestProject")
+            public class FileInterceptorsDeclaration {
+                @InterceptCalls
+                @InstanceMethod
+                @InterceptInherited
+                public static String intercept_getDescription(@Receiver Rule thisRule) {
+                    return "";
+                }
+            }
+        """
+        def propertyUpgrade = source """
+            package org.gradle.test;
+
+            import org.gradle.api.provider.Property;
+            import org.gradle.internal.instrumentation.api.annotations.VisitForInstrumentation;
+            import org.gradle.internal.instrumentation.api.annotations.UpgradedProperty;
+
+            public abstract class Task {
```

Comment:
:x: Let's not use `Task` as a type name here as it's easy to confuse with our API type. `CustomTask` should be a good alternative.